File: question/neetcode150/backtracking/combination-sum.py
from typing import List

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []

        s = []

        def dfs(remaining, i):
            if remaining == 0:
                ans.append(s.copy())
                return
            if i == len(candidates):
                return
            
            if remaining < 0:
                return 

            s.append(candidates[i])
            dfs(remaining-candidates[i], i)
            s.pop()

            dfs(remaining,i+1)

        dfs(target,0)
        return ans

File: question/neetcode/backtracking/combination-sum.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
    public:
        void dfs(int i, int s, int n, vector<int>& cur, vector<vector<int>>& ans, vector<int>& candidates, int target){
            if(s==target){
                ans.push_back(cur);
                return;
            }
            
            if(i==n or s>target){
                return;
            }

            cur.push_back(candidates[i]);
            dfs(i,s+candidates[i],n,cur,ans,candidates,target);
            cur.pop_back();
            dfs(i+1,s,n,cur,ans,candidates,target);
        }

        vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
            vector<int> cur;
            vector<vector<int>> ans;
            int n = candidates.size();

            dfs(0,0,n,cur,ans,candidates,target);

            return ans;
        }
    };